python/kfserving/kfserving/storage.py

```diff
@@ -43,15 +47,48 @@ def download(uri: str) -> str:
 
     @staticmethod
     def _download_s3(uri, temp_dir: str):
-        raise NotImplementedError
+        client = Storage._create_minio_client()
+        bucket_args = uri.replace(_S3_PREFIX, "", 1).split("/", 1)
+        bucket_name = bucket_args[0]
+        bucket_path = bucket_args[1] if len(bucket_args) > 1 else ""
+        objects = client.list_objects(bucket_name, prefix=bucket_path, recursive=True)
+        for obj in objects:
+            object_file_name = obj.object_name.replace(bucket_path, "", 1).strip("/")
+            client.fget_object(bucket_name, obj.object_name, temp_dir + "/" + object_file_name)
 
     @staticmethod
     def _download_gcs(uri, temp_dir: str):
-        raise NotImplementedError
+        try:
+            storage_client = storage.Client()
+        except exceptions.DefaultCredentialsError as e:
+            storage_client = storage.Client.create_anonymous_client()
+        bucket_args = uri.replace(_GCS_PREFIX, "", 1).split("/", 1)
+        bucket_name = bucket_args[0]
+        bucket_path = bucket_args[1] if len(bucket_args) > 1 else ""
+        bucket = storage_client.bucket(bucket_name)
+        blobs = bucket.list_blobs(prefix=bucket_path)
+        for blob in blobs:
+            object_file_name = blob.name.replace(bucket_path, "", 1).strip("/")
+            # Create necessary subdirectory
```

Comment:
Can we use exist_ok=True
https://docs.python.org/3/library/os.html#os.makedirs

Reply:
done